akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -142,6 +142,62 @@ private[akka] final class HeadOptionSink[In](val attributes: Attributes, shape:
 
 /**
  * INTERNAL API
+ */
+private[akka] object LastSink {
+  final class LastOptionSinkSubscriber[In] extends Subscriber[In] {
+    private[this] var prev: In = null.asInstanceOf[In]
+    private[this] var subscription: Subscription = null
+    private[this] val promise: Promise[Option[In]] = Promise[Option[In]]()
+    def future: Future[Option[In]] = promise.future
+    override def onSubscribe(s: Subscription): Unit = {
+      ReactiveStreamsCompliance.requireNonNullSubscription(s)
+      if (subscription ne null) s.cancel()
+      else {
+        subscription = s
+        s.request(Long.MaxValue)
+      }
+    }
+
+    override def onNext(elem: In): Unit = {
+      ReactiveStreamsCompliance.requireNonNullElement(elem)
+      prev = elem
+    }
+
+    override def onError(t: Throwable): Unit = {
+      ReactiveStreamsCompliance.requireNonNullException(t)
+      promise.tryFailure(t)
+      subscription = null
+    }
+
+    override def onComplete(): Unit = {
+      promise.trySuccess(Option(prev))
+      prev = null.asInstanceOf[In]
+      subscription = null
+    }
+  }
```

Comment:
instead of adding this (and applying the TCK) I’d strongly prefer a GraphStage implementation